File: services/news_api/app/models/news_item.py
"""News item model."""

from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class NewsItem(Base):
    __tablename__ = "news_items"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    category_id: Mapped[int] = mapped_column(Integer, ForeignKey("categories.id"), index=True)
    source_id: Mapped[int] = mapped_column(Integer, ForeignKey("sources.id"), index=True)
    title: Mapped[str] = mapped_column(String(500))
    summary: Mapped[str] = mapped_column(Text, default="")
    url: Mapped[str] = mapped_column(String(1200), index=True)
    cover_image: Mapped[str] = mapped_column(String(1200), default="")
    published_at: Mapped[datetime] = mapped_column(DateTime, index=True)
    fetched_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, index=True)
    tags: Mapped[str] = mapped_column(String(500), default="[]")
    score: Mapped[int] = mapped_column(Integer, default=0)
    dedup_hash: Mapped[str] = mapped_column(String(64), unique=True, index=True)

File: apps/news_app/lib/core/models/news_item.dart
class NewsItemModel {
  NewsItemModel({
    required this.id,
    required this.categoryId,
    required this.sourceId,
    required this.title,
    required this.summary,
    required this.url,
    required this.coverImage,
    required this.publishedAt,
    required this.fetchedAt,
    required this.tags,
    required this.score,
  });

  factory NewsItemModel.fromJson(Map<String, dynamic> json) {
    return NewsItemModel(
      id: json['id'] as int,
      categoryId: json['category_id'] as int,
      sourceId: json['source_id'] as int,
      title: json['title'] as String? ?? '',
      summary: json['summary'] as String? ?? '',
      url: json['url'] as String? ?? '',
      coverImage: json['cover_image'] as String? ?? '',
      publishedAt:
          DateTime.tryParse(json['published_at'] as String? ?? '') ??
          DateTime.now(),
      fetchedAt:
          DateTime.tryParse(json['fetched_at'] as String? ?? '') ??
          DateTime.now(),
      tags: json['tags'] as String? ?? '[]',
      score: json['score'] as int? ?? 0,
    );
  }

  final int id;
  final int categoryId;
  final int sourceId;
  final String title;
  final String summary;
  final String url;
  final String coverImage;
  final DateTime publishedAt;
  final DateTime fetchedAt;
  final String tags;
  final int score;
}
